rspamc: add -R option for human readable report

amishmm opened the pull request:
Spamassassin has a nice feature of printing human readable report. (see spamc -R option)

I had created a feature request for the same which describes the requirement in detail. The feature request can be read here #4373 

This pull request (PR) tries to emulate "spamc -R" option of spamassassin.

First line will print score/threshold,Action=action,Spam=bool and then followed by human readable report.

Sample report is as follows:

```
10.80/15.00,Action=add header,Spam=true
Content analysis details:   (10.80 points, 15.00 required)

 pts rule name              description
---- ---------------------- --------------------------------------------------
 0.0 ARC_NA                 ARC signature absent
 0.0 ASN                    [asn:34245, ipnet:212.17.32.0/19, country:IE]
 0.1 DMARC_POLICY_SOFTFAIL  DMARC failed [buap.mx : No valid SPF, No valid DKI
                            M,none]
 0.0 FROM_NEQ_ENVFROM       From address is different to the envelope [xl6Ety0
                            0V@fismat1.fcfm.buap.mx,dev_null_sample_spam@examp
                            le.com]
 0.0 FROM_NO_DN             From header does not have a display name
 1.5 INVALID_DATE           Malformed date header
 1.7 INVALID_MSGID          Message id is incorrect
-0.1 MIME_GOOD              Known content-type [text/plain]
 0.0 MIME_TRACE             [0:+]
 0.0 RCPT_COUNT_ONE         One recipient [1]
 0.0 RCVD_COUNT_FIVE        Message has 5-7 Received headers [6]
 0.1 RCVD_NO_TLS_LAST       Last hop did not use encrypted transports
 0.0 R_DKIM_NA              Missing DKIM signature
 1.0 R_SPF_FAIL             SPF verification failed [-all:c]
 3.0 R_UNDISC_RCPT          Recipients are absent or undisclosed
 0.0 TO_DN_ALL              All the recipients have display names
 3.5 VIOLATED_DIRECT_SPF    Has no Received (or no trusted received relays) an
                            d SPF policy fails or soft fails

Emails: ["maillistdrop@post.com"]

```

This is more or less same output that spamc of spamassassin gives with its -R option.

I have done some more additions to it, like printing of URLS and Emails (if available) at the end. Which can act as a preview for links embedded in the email.

I have tried to do minimal changes to the existing "rspamc_symbols output()" function. Such that it does not change the original flow of the code.

Please review the PR.

Thank you

vstakhov commented:
It looks veru good in general, thank you! I will review details a bit later.

vstakhov commented on src/client/rspamc.cxx:
```diff
@@ -823,6 +825,59 @@ add_options(GQueue *opts)
+{
+	if (maxlen < 1) return;
+
+	std::string s;
+	for (size_t pos = 0; pos < line.length(); pos += s.length()) {
+		s = line.substr(pos, pos ? (maxlen-indent) : maxlen);
+		if (indent && pos) fmt::print(out, "{:>{}}", " ", indent);
```
```suggestion
	for (size_t pos = 0; pos < line.length(); pos += s.length()) {
		auto s = std::string_view(line).substr(pos, pos ? (maxlen-indent) : maxlen);
		if (indent && pos) {
			fmt::print(out, "{:>{}}", " ", indent);
		}
```

amishmm replied:
Can't define "s" as "auto" inside the "for" loop as otherwise "s.length()" becomes outside the scope and compiler throws the error.

vstakhov commented on src/client/rspamc.cxx:
```diff
@@ -823,6 +825,59 @@ add_options(GQueue *opts)
+static void
+print_indented_line(FILE *out, std::string line, size_t maxlen, size_t indent)
+{
+	if (maxlen < 1) return;
```
```suggestion
	if (maxlen < 1) {
		return;
	}
```

vstakhov commented on src/client/rspamc.cxx:
```diff
@@ -823,6 +825,59 @@ add_options(GQueue *opts)
+rspamc_symbol_human_output(FILE *out, const ucl_object_t *obj)
+{
+	auto first = true;
+	double score = 0;
```
```suggestion
	auto score = 0.0;
```

vstakhov commented on src/client/rspamc.cxx:
```diff
@@ -823,6 +825,59 @@ add_options(GQueue *opts)
+	const auto *val = ucl_object_lookup(obj, "score");
+	if (val != nullptr) score = ucl_object_todouble(val);
+
+	key = ucl_object_key(obj);
```
```suggestion
	const auto *key = ucl_object_key(obj);
```

vstakhov commented on src/client/rspamc.cxx:
```diff
@@ -823,6 +825,59 @@ add_options(GQueue *opts)
+{
+	auto first = true;
+	double score = 0;
+	const char *key = nullptr, *desc = nullptr;
```
```suggestion
	const char *desc = nullptr;
```

vstakhov commented on src/client/rspamc.cxx:
```diff
@@ -823,6 +825,59 @@ add_options(GQueue *opts)
+	val = ucl_object_lookup(obj, "description");
+	if (val != nullptr) desc = ucl_object_tostring(val);
+
+	std::string line = fmt::format("{:>4.1f} {:<22} ", score, key);
```
```suggestion
	auto line = fmt::format("{:>4.1f} {:<22} ", score, key);
```

vstakhov commented on src/client/rspamc.cxx:
```diff
@@ -823,6 +825,59 @@ add_options(GQueue *opts)
+	if (desc != nullptr) line += desc;
+
+	val = ucl_object_lookup(obj, "options");
+	if (val != nullptr && val->type == UCL_ARRAY) {
```
```suggestion
	if (val != nullptr && ucl_object_type(val) == UCL_ARRAY) {
```

amishmm replied:
This line was copied from rspamc_symbol_output() - so I have made above changes to this line as well as the line which was copied.

vstakhov commented on src/client/rspamc.cxx:
```diff
@@ -864,12 +919,12 @@ rspamc_metric_output(FILE *out, const ucl_object_t *obj)
 		auto *elt = ucl_object_lookup(obj, ucl_name);
 		if (elt) {
-			fmt::print(out, "{}: {}\n", output_message,
+			fmt::print(out, fmt::runtime(humanreport ? ",{}={}" : "{}: {}\n"), output_message,
 					emphasis_argument(ucl_object_tostring(elt)));
```
```suggestion
			if (humanreport) {
				fmt::print(out, {}={}", output_message, emphasis_argument(ucl_object_tostring(elt)));
			}
			else {
				fmt::print(out, "{}: {}\n", output_message,
					emphasis_argument(ucl_object_tostring(elt)));
			}
```

vstakhov commented on src/client/rspamc.cxx:
```diff
@@ -1003,7 +1079,7 @@ rspamc_symbols_output(FILE *out, ucl_object_t *obj)
 		}
 
-		fmt::print(out, "Urls: {}\n", emitted);
+		if (emitted && strcmp(emitted, "[]")) fmt::print(out, "Urls: {}\n", emitted);
```
This will change the existing behaviour which is not a good thing. I'd suggest to use this logic merely for reports.

amishmm replied:
I did this intentionally. But noted and will change it back. Thanks.

vstakhov commented on src/client/rspamc.cxx:
```diff
@@ -915,26 +977,36 @@ rspamc_metric_output(FILE *out, const ucl_object_t *obj)
 
-			fmt::print(out, "Spam: {}\n", emphasis_argument(act.value() < METRIC_ACTION_GREYLIST ?
-											  "true" : "false"));
+			fmt::print(out, fmt::runtime(humanreport ? ",Spam={}" : "Spam: {}\n"),
```
I'd suggest to convert `fmt::runtime` to a simple `if...else` to keep fmt constexpr providing better compile time checks.

vstakhov commented on src/client/rspamc.cxx:
```diff
@@ -915,26 +977,36 @@ rspamc_metric_output(FILE *out, const ucl_object_t *obj)
 					break;
 				}
-				fmt::print(out, "Action: {}\n", colorized_action);
+				fmt::print(out, fmt::runtime(humanreport ? ",Action={}" : "Action: {}\n"), colorized_action);
```
Same, no `fmt::runtime` please.

vstakhov commented on src/client/rspamc.cxx:
```diff
@@ -1018,11 +1094,12 @@ rspamc_symbols_output(FILE *out, ucl_object_t *obj)
 		}
 
-		fmt::print(out, "Emails: {}\n", emitted);
+		if (emitted && strcmp(emitted, "[]")) fmt::print(out, "Emails: {}\n", emitted);
```
Same here, please let it be `humanreport` specific behaviour.

vstakhov reviewed (approved):
Few nits added.

amishmm commented:
I have updated the PR, implementing the suggestions given by reviewer @vstakhov .

There are new additions as follow:

1. The first line of the output contains the message score and three threshold scores, in this format:<br>
score/greylist/addheader/reject,action=N:ACTION,spam=0|1,skipped=0|1<br>
This is done because rspamd is bit different from spamassassin and uses 3 types of score thresholds.<br>
This allows to generate statistics of messages greylisted or detected as spam or rejected based on the score alone without using action or spam fields of the first line.<br>
action=N:ACTION, here N is the number indicating the action type. ACTION is the textual desciption of the same.<br>
NOTE: Currently rspamd does not provide "greylist_score" and "addheader_score" in JSON / UCL reply. But I have added the same in rspamc code.<br>
When I have time, I will try to look at the rspamd code and implement those scores.

2. Emails and Urls line now respect 78 character limit (for human report) 

Please review the code again.

Thank you

amishmm commented:
Thank you very much for the merge.

Just a suggestion, its better to squash and merge as a single commit so that the "trial / incomplete / test" commits don't go in master branch but only the final approved code goes to master branch.
